packages/types/task/index.ts

```diff
@@ -0,0 +1,41 @@
+import { EventEmitter } from 'eventemitter3';
+import Step from '../step';
+import Workflow from '../workflow';
+
+interface TaskStepEventData<T extends Task>  {
+  task: T;
+  step: Step;
+}
+
+interface TaskEventData<T extends Task> {
+  task: T;
+}
+
+// Task is a class I don't see much need for custom implementations
+class TaskImpl extends EventEmitter implements Task {
+  public id: string;
+  public assigneeId: string;
+  public watcherIds: string[];
+  public workflowId: string;
+  public currentStep: Step;
+  public steps: Step[];
+}
+
+
+interface Task extends EventEmitter {
+  id: string;
+  assigneeId: string;
+  // multiple Users might want to watch a Task's progression
+  watcherIds: string[];
+
+  // If we take a snapshot of the Workflow's Steps we might not need a reference to it
+  // maybe just for keeping a historical/UI link
+  // workflowId: string;
+
+  steps: Step[];
+  currentStep: Step;
+
+  on(event: 'step:change', handler: (e: TaskStepEventData<this>) => any): this;
```

Comment:
I think events are a pretty well established way of having hooks into actual 'events' that are important for client code. The only alternative to hooking into those lifecycle steps would be to actually extend or wrap the implementation with the extra functionality, or supply them as custom callback-receiving functions, which is essentially the same thing as an event emitter.

The difference to the mediator pattern is that they are not intended to be used for inter-module communications and method calling, only as ways to observe specific parts of the lifecycle.